src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -317,6 +258,61 @@ impl<T: Timestamp + Lattice + Codec64 + Display + Sync> AsyncStorageWorker<T> {
+                        // The as-of frontier controls the frontier with which all inputs of the
+                        // ingestion dataflow will be advanced by. It is in our interest to set the
+                        // as-of froniter to the largest possible value, which will result in the
+                        // maximum amount of consolidation, which in turn results in the minimum
+                        // amount of memory required to hydrate.
+                        //
+                        // For each output `o` and for each input `i` of the ingestion the
+                        // controller guarantees that i.since < o.upper except when o.upper is
+                        // [T::minimum()]. Therefore the largest as-of for a particular output `o`
+                        // is `{ (t - 1).advance_by(i.since) | t in o.upper }`.
+                        //
+                        // To calculate the global as_of frontier we take the minimum of all those
+                        // per-output as-of frontiers.
+                        let client = persist_clients
+                            .open(
+                                ingestion_description
+                                    .ingestion_metadata
+                                    .persist_location
+                                    .clone(),
+                            )
+                            .await
+                            .expect("error creating persist client");
+                        let read_handle = client
+                            .open_leased_reader::<SourceData, (), T, Diff>(
+                                seen_remap_shard,
+                                Arc::new(ingestion_description.desc.connection.timestamp_desc()),
+                                Arc::new(UnitSchema),
+                                Diagnostics {
+                                    shard_name: ingestion_description
+                                        .remap_collection_id
+                                        .to_string(),
+                                    handle_purpose: format!("resumption data for {}", id),
+                                },
+                                false,
+                            )
+                            .await
+                            .unwrap();
+                        let remap_since = read_handle.since().clone();
+                        mz_ore::task::spawn(move || "deferred_expire", async move {
+                            tokio::time::sleep(std::time::Duration::from_secs(300)).await;
+                            read_handle.expire().await;
+                        });
+                        let mut as_of = Antichain::new();
+                        for upper in resume_uppers.values() {
+                            for t in upper.elements() {
+                                let mut t_prime = t.step_back().unwrap_or(T::minimum());
+                                if !remap_since.is_empty() {
```

Comment:
This special case could use a comment! I think an empty since here means that the ingestion has already been dropped so we should return `as_of = []`? Would be good to spell that out a bit.

Reply:
will do